### PR TITLE
[Merged by Bors] - Fix unreachable panics in compile_access

### PR DESCRIPTION
**Pull request by VTCAKAVSMoACE**
This PR changes the following:

- More elegantly handles illegal access statements in compile_access
- Adds a slew of previously unhandled illegal access test cases

### Caveats

It is very, very likely that you will want to simply restrict unary and assignment operations in the AST. However, this prevents crashes in the meantime with a error that is just slightly less detailed than if it were implemented in AST.

### CODE REVIEW

**Review by Razican — APPROVED** [review]
Looks good to me!

File: boa_engine/src/tests.rs [hunk 1]
```diff
@@ -1357,6 +1405,25 @@ fn multicharacter_bitwise_assignment_to_non_assignable() {
...
+    // Disabled - awaiting implementation.
+    let test_cases = [
+        "(()=>{})() >>>= 5",
+        "(()=>{})() &&= 5",
+        "(()=>{})() ||= 5",
+        "(()=>{})() ??= 5",
+    ];
+
+    for case in &test_cases {
+        let string = dbg!(forward(&mut context, case));
```
Comment (jedel1043) [inline boa_engine/src/tests.rs:1421]:
```suggestion
        let string = forward(&mut context, case);
```

File: boa_engine/src/tests.rs [hunk 2]
```diff
@@ -1342,6 +1377,19 @@ fn multicharacter_assignment_to_non_assignable() {
...
 }
 
+#[test]
+fn multicharacter_assignment_to_non_assignable_ctd() {
+    let mut context = Context::default();
+
+    let test_cases = ["(()=>{})() **= 5", "(()=>{})() <<= 5", "(()=>{})() >>= 5"];
+
+    for case in &test_cases {
+        let string = dbg!(forward(&mut context, case));
```
Comment (jedel1043) [inline boa_engine/src/tests.rs:1387]:
```suggestion
        let string = forward(&mut context, case);
```

File: boa_engine/src/tests.rs [hunk 3]
```diff
@@ -485,6 +485,18 @@ fn unary_pre() {
     assert_eq!(&exec(execs_before_dec), "true");
 }
 
+#[test]
+fn invalid_unary_access() {
+    let mut context = Context::default();
+    let test_cases = ["++[];", "[]++;", "--[];", "[]--;"];
+
+    for case in &test_cases {
+        let string = dbg!(forward(&mut context, case));
```
Comment (jedel1043) [inline boa_engine/src/tests.rs:494]:
```suggestion
        let string = forward(&mut context, case);
```

**Review by jedel1043 — CHANGES_REQUESTED** [review]
Good job! Just some `dbg` leftovers

File: boa_engine/src/tests.rs [hunk 4]
```diff
@@ -1326,6 +1338,29 @@ fn assignment_to_non_assignable() {
...
+        "(()=>{})() |= 5",
+        "(()=>{})() += 5",
+        "(()=>{})() = 5",
+    ];
+
+    for case in &test_cases {
+        let string = forward(&mut context, case);
+
+        assert!(string.starts_with("Uncaught \"SyntaxError\": "));
+    }
```
Comment (HalidOdat) [inline boa_engine/src/tests.rs:1361]: We have a better way of testing in rust with https://github.com/boa-dev/boa/blob/main/boa_engine/src/lib.rs#L151-L193
Here is an example https://github.com/boa-dev/boa/blob/main/boa_engine/src/tests.rs#L1360-L1368

**Comment by HalidOdat** [comment]
> ### Caveats
> 
> It is very, very likely that you will want to simply restrict unary and assignment operations in the AST. However, this prevents crashes in the meantime with a error that is just slightly less detailed than if it were implemented in AST.

Yup. These are early errors that should be checked during parsing, and the `unreachable!()` should be unreachable if we implement them.

File: boa_engine/src/tests.rs (hunk 4 above)
Comment (VTCAKAVSMoACE) [reply boa_engine/src/tests.rs:1361]: Okay, I'll change that over; I was following the pattern I observed with similar test cases.

File: boa_engine/src/bytecompiler.rs [hunk 5]
```diff
@@ -456,13 +456,13 @@ impl<'b> ByteCompiler<'b> {
...
-            Node::Identifier(name) => Access::Variable { name: name.sym() },
-            Node::GetConstField(node) => Access::ByName { node },
-            Node::GetField(node) => Access::ByValue { node },
-            Node::This => Access::This,
-            _ => unreachable!(),
+            Node::Identifier(name) => Some(Access::Variable { name: name.sym() }),
+            Node::GetConstField(node) => Some(Access::ByName { node }),
+            Node::GetField(node) => Some(Access::ByValue { node }),
+            Node::This => Some(Access::This),
+            _ => None,
```
Comment (jedel1043) [inline boa_engine/src/bytecompiler.rs:465]: This Clippy warning was already fixed, but the merge undid the fix

**Review by jedel1043 — CHANGES_REQUESTED** [review]

File: boa_engine/src/bytecompiler.rs (hunk 5 above)
Comment (VTCAKAVSMoACE) [reply boa_engine/src/bytecompiler.rs:465]: Ah hell, sorry. Fixing.

**Review by jedel1043 — APPROVED** [review]
Looks good to me!

**Pull request closed without merging**